### ISSUE TITLE
Upgrading large databases (>400k posts) results in statement timeouts on PostgreSQL

### ISSUE THREAD

**Issue by Diftraku** [shish/shimmie2#847]
**Server Software**
(You can get all these stats from `http://<your site>/system_info`)

 - Shimmie version: git tag v2.9.1 (old database is from git tag v2.8.4)
 - Database: PostgreSQL 13.2
 - Web server: nginx 1.21.1

**Client Software (please complete the following information)**
 - Device: Linux desktop / terminal
 - Browser: Google Chrome Version 93.0.4577.82 (Official Build) (64-bit)

**What steps trigger this bug**
1. Prepare a copy of the code at tag `v2.9.1`
2. Prepare a database with large number of posts (>400k) of varying types, using `v2.8.4`  schema
3. Trigger an upgrade, using one of the following methods
3a. Navigate to the front page of the Shimmie install
3b. Run `php index.php` via terminal
4. Wait for a couple of seconds
5. Transaction timeout occurrs and upgrade is rolled back to last completed transaction.

**What did you expect to happen?**
Database upgrade proceeds as normal, taking time to process the large number of posts and applying media related changes.

**What actually happened?**
From the command-line:
```
root@app-f45b4fcf8-hjclg:/app# php index.php 
2021-09-22T00:33:30+03:00 upgrade: Setting predictable media values for known file types
PHP Notice:  Undefined index: args in /app/core/util.php on line 606
PHP Warning:  array_map(): Expected parameter 2 to be an array, null given in /app/core/util.php on line 606
PHP Warning:  implode(): Invalid arguments passed in /app/core/util.php on line 606
PHP Notice:  Undefined index: args in /app/core/util.php on line 606
PHP Warning:  array_map(): Expected parameter 2 to be an array, null given in /app/core/util.php on line 606
PHP Warning:  implode(): Invalid arguments passed in /app/core/util.php on line 606
PHP Notice:  Undefined index: args in /app/core/util.php on line 606
PHP Warning:  array_map(): Expected parameter 2 to be an array, null given in /app/core/util.php on line 606
PHP Warning:  implode(): Invalid arguments passed in /app/core/util.php on line 606
Trace: 0: /app/index.php(72): send_event()
1: /app/core/send_event.php(113): Upgrade->onDatabaseUpgrade()
2: /app/ext/upgrade/main.php(196): Database->execute()
Message: SQLSTATE[57014]: Query canceled: 7 ERROR:  canceling statement due to statement timeout
Version: 2.9.1+ (on 7.4.21)
```
From browser:
![image](https://user-images.githubusercontent.com/171226/134253445-770ef443-e55f-491e-bc94-3a81af5839f7.png)

Checking the global config on the PostgreSQL server for `statement_timeout` shows it's set to 0 (disabled, as per [docs](https://www.postgresql.org/docs/13/runtime-config-client.html#:~:text=taken%20as%20milliseconds.-,A%20value%20of%20zero%20(the%20default)%20disables%20the%20timeout.,-The%20timeout%20is))

---

The first hurdle is when `upgrade` is updating some flags related to JPEG images: https://github.com/shish/shimmie2/blob/v2.9.1/ext/upgrade/main.php#L196

The sheer amount of data to chug through here results in a timeout (somewhere) to be exceeded. To get going, we throw a shit ton of excessively large timeouts between statements and get all the way to bumping `db_version` to 21. 

Switching to `media` extension, we get to bumping `ext_media_version` to 4 and have our next timeout:
This time, it's the `media` extension trying to do similar changes as `upgrade` did before (handling image related field changes): https://github.com/shish/shimmie2/blob/v2.9.1/ext/media/main.php#L925

Sorting the timeouts lets the upgrade continue as usual.

It might be worth-while to either consistently call `Database::set_timeout()` during upgrades, or in case of PostgreSQL as the database: add a unit to the timeout or outright disable it.

From: https://www.postgresql.org/docs/13/runtime-config-client.html#:~:text=statement_timeout%20(integer)
> Abort any statement that takes more than the specified amount of time. If log_min_error_statement is set to ERROR or lower, the statement that timed out will also be logged. If this value is specified without units, it is taken as milliseconds. A value of zero (the default) disables the timeout.

**Comment by Diftraku** [comment]
Once I got past the upgrades, I had issues even loading the front page and opted to add a kludge in `Database::set_timeout()` so I can at least browse the site:
```php
    public function set_timeout(int $time): void
    {
        if ($db_proto === DatabaseDriver::PGSQL) {
            // Make the timeout a bit more sane
            if ($time >= 300000) {
                $time = $time / 1000;
            }
            $time = $time.'s';
        }
        $this->engine->set_timeout($this->db, $time);
    }
```

**Comment by shish** [comment]
Currently some of the heavier database code assumes low timeouts, and sets the timeout to a "high" value (5 minutes), which is probably why you'd see timeouts even with the default being "disabled"? I'm putting up a commit now which should change that code to disable timeouts instead of setting it to 5 minutes though :)

Also maybe relevant is `SPEED_HAX` mode which does a bunch of things, one of which is disabling auto-upgrades (so that if you have a bunch of users, they don't all trigger the upgrade code in parallel), then you manually kick off upgrades via CLI -- https://github.com/shish/shimmie2/wiki/Performance -- doing it via CLI also means that you don't risk the upgrade getting interrupted by an HTTP timeout.

> add a kludge in `Database::set_timeout()`

If it's postgres specific, it could go in `PostgreSQL::set_timeout()` and no need to check `$db_proto`.  I'm not sure what practical effect changing from milliseconds to seconds has though o.o

**Comment by Diftraku** [comment]
> I'm putting up a commit now which should change that code to disable timeouts instead of setting it to 5 minutes though :)

Super, I have a hunch the random timeouts for normal pageloads were just stuff being IO-bound as I haven't had it happening after the upgrades were done and the container running the code with FPM was restarted a few times. Nothing a few `kubectl rollout restart`s won't fix :3

> Also maybe relevant is SPEED_HAX mode which does a bunch of things, one of which is disabling auto-upgrades

Definitely something I'll look into, at least for the upgrades (enabling it before actually updating anything on the install).

> If it's postgres specific, it could go in PostgreSQL::set_timeout()

Completely blanked out on this, it seems. I did a very cursory glance through `dbengine.php` but likely missed it while trying to hack away at the wee hours of the morning. I did get a good laugh from spotting the name of the package where the PDO was being imported from in `database.php`.

> I'm not sure what practical effect changing from milliseconds to seconds has though o.o

In my defense, it seemed like a lot greater idea than it actually was at 2AM xD

I probably thought there were small timeouts littered throughout the code, after having issues getting the frontpage from `home` extension to load (getting a timeout instead of the counters).

Given that everything timed out after 5-7 seconds during the upgrades and I didn't want to just tack on the unit and have something that used to be for 5 minutes suddenly turn into a waiting game of 4 hours (plus change), hence the arbitrary division shenanigans.

Basically, every upgrade that touched the `images` table caused a timeout to happen awfully quickly, leading me to believe `PostgreSQL::execute()` (or something else) kept resetting the timeout between individual calls.

**Comment by Diftraku** [comment]
I finally got around to testing my original dump (db_version = 18) against 3061a9d7d5d1e73b3ea4c955aa4b6efd18f00153 *without* `SPEED_HAX` enabled. Both browser and CLI end the upgrade early due to a query timeout, as is probably expected to prevent DoS'ing the site via database upgrades.

Enabling `SPEED_HAX` will apply the query timeout correctly and the upgrade completes without issues (apart from my PSQL instance choking due to resource constraints). The most heaviest change seems to be with DB version 21 (adding mime types).

While not exactly the same version as where the fix was originally introduced in the CLI (3bb1566df28067bf6cadf553b1f5ae3f13534231), I can confirm the fix does indeed work.
